extra_data/keydata.py

```diff
@@ -340,13 +340,38 @@ def as_single_value(self, rtol=1e-5, atol=0.0, reduce_by='median'):
         the first value encountered. By default, 'median' is used.
 
         If within tolerances, the reduced value is returned.
+
+        For non-numerical keys like strings, the method instead always
+        checks for uniqueness and returns such a value, if present.
         """
 
         data = self.ndarray()
 
         if len(data) == 0:
             raise NoDataError(self.source, self.key)
 
+        if not np.issubdtype(self.dtype, np.number):
+            # Handle non-numeric types first.
+
+            if reduce_by is not None:
+                raise ValueError('custom reduce method not supported for '
+                                 'non-numeric type')
+
+            unique_values = np.unique(data, axis=None)
+
+            if len(unique_values) > 1:
+                raise ValueError(f'str values are not unique: {unique_values}')
+
+            value = unique_values[0]
+
+            if isinstance(value, bytes):
+                value = value.decode('utf-8', 'surrogateescape')
```

Comment:
I'd lean towards skipping this so that `kd.as_single_value()` gives us the same type as `kd.ndarray()[0]`. I don't feel strongly about it, though - I accept people are probably more likely to want strings, I just like consistency.

Reply:
Ah, as mentioned above I would quite like to expose strings as strings. As there's been no type from the DAQ before that could end up in `bytes` in files, it seems safe to convert. I agree with your point of consistency, so how about also converting it in the other getters?

Reply:
Right, sorry, I'd forgotten our discussion about that. That seems reasonable. If we can get numpy 2.x in our environments we could use the new numpy variable length string type, which is meant to be more efficient. But the new 202502 environment still has Numpy 1.x, so maybe that's overambitious. Hopefully string data should all be relatively small anyway.

Reply:
Ah right, that is tricky for arrays. So we just go with fixed-length strings conversion? What should `KeyData.dtype` report in such a case? 🤔

Reply:
I'd guess it might be more efficient to do the conversion in one go rather than making an array of fixed width strings as an intermediate. This is what h5py does:

https://github.com/h5py/h5py/blob/6f0d8f0d411eacf8ac6da90302996e2fe9a95619/h5py/_hl/dataset.py#L292-L294

I believe most of the performance advantage of numpy over Python operations goes out the window once object arrays are involved. I haven't really tested this, though.